### ISSUE TITLE
pip interop for relative packages

### ISSUE THREAD

**Issue by serjflint** [astral-sh/uv#2635]
Hi! 

FIrst, I wish to say that uv is an amazing tool and works blazingly fast.

I am working with an existing legacy setup and trying to use uv as a local replacement for pip. 
The biggest complication with that legacy setup that it requires a package from a local relative path.
I have to use https://peps.python.org/pep-0440/#direct-references for it.

In the issues I have found that I can use `${PROJECT_ROOT}` for the uv. But it is not really an environment variable and it is not set at the time of loading 'setup.py' file.

In the end I got something like this

```python
import os
import pathlib
import setuptools
import sys

PACKAGE_PATH = (
    pathlib.Path(__file__).parent.parent / 'dir' / 'package'
)
if sys.argv[0] == '-c':  # uv hack
    PACKAGE_URL = r'file:///${PROJECT_ROOT}/../dir/package'
else:
    PACKAGE_URL = f'file://localhost/{PACKAGE_PATH}#egg=sub_package'
```

Is there any better way to do this?

**Comment by serjflint** [comment]
When I used `CODE_LINTERS_URL = r'file:///${PROJECT_ROOT}/../github/code-linters#egg=taxi_linters'`
and called `python -m pip install -e .` I got a following error:
`ERROR: Could not install packages due to an OSError: [Errno 2] No such file or directory: '/github/code-linters'`

**Comment by charliermarsh** [comment]
We accept relative URLs for direct references, so you can do like `foo @ ./bar`. Does that work?

**Comment by serjflint** [comment]
> We accept relative URLs for direct references, so you can do like `foo @ ./bar`. Does that work?

I tried `package @ ../dir/package` and got an error with pip 

`error in main-package setup command: 'install_requires' must be a string or list of strings containing valid project/version requirement specifiers; Invalid URL: ../dir/package`

The question is how to get a configuration which works with both pip and uv.

**Comment by charliermarsh** [comment]
Yeah, unfortunately `pip` doesn't support relative references. It does expand environment variables, so you _could_ use `PROJECT_ROOT` like above, but you'd have to "inject" it manually when running `pip install` commands.

**Comment by serjflint** [comment]
When I use `f'file://localhost/{PACKAGE_PATH}#egg=sub_package'` I get an error

```bash
uv pip install -e .
error: Failed to build editables
  Caused by: Failed to parse metadata from built wheel
  Caused by: relative path without a working directory: localhost//abs/path/dir/package#egg=sub_package
yandex-taxi-code-linters @ file://localhost//abs/path/dir/package#egg=sub_package
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by serjflint** [comment]
I assume it is wrongly treated as a relative path.

**Comment by charliermarsh** [comment]
Why include `localhost/`?

**Comment by serjflint** [comment]
> Why include `localhost/`?

It is an assumed default by pep 440 https://peps.python.org/pep-0440/#file-urls

Without it - `PACKAGE_URL = f'file:///{PACKAGE_PATH}#egg=sub_package'` I got an error

```bash
error in main-package setup command: 'install_requires' must be a string or list of strings containing valid project/version requirement specifiers; Invalid URL given
```

**Comment by charliermarsh** [comment]
Thanks, we're probably just missing handling for that, although I think the problem is your `PACKAGE_PATH` has a leading slash...? The resolved URL was `file://localhost//abs/path/dir/package#egg=sub_package`.

**Comment by serjflint** [comment]
> Thanks, we're probably just missing handling for that, although I think the problem is your `PACKAGE_PATH` has a leading slash...? The resolved URL was `file://localhost//abs/path/dir/package#egg=sub_package`.

Actually yes. I removed it and it solved the problem. Thank you very much for your time and patience.

**Comment by charliermarsh** [comment]
I filed an issue for the `localhost` part here: https://github.com/astral-sh/uv/issues/2652

**Comment by bluss** [comment]
Now I don't know your project, but in general about ${PROJECT_ROOT}, there are package systems that support that variable and can create packages using it that pip can install, and setuptools is not one of them.
PDM seems to be one of them. In such a PDM defined pyproject, pip reads the project, starts the wheel build, pdm build hooks resolve the dependency to an absolute path and it somehow works.

The general thinking would be to use a package builder that supports relative path references like that. But I'll be careful and say I don't know which way works simultaneously in both pip and uv. The way I would currently do that works in pip but falls afoul of #1808 in uv (for a file URL).
